libraries/AP_BattMonitor/AP_BattMonitor_TattuCAN.cpp

```diff
@@ -0,0 +1,158 @@
+/*
+ * This file is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This file is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Author: Tom Pittenger
+ */
+
+#include "AP_BattMonitor_TattuCAN.h"
+
+#if AP_BATT_MONITOR_TATTU_ENABLED
+#include <AP_HAL/AP_HAL.h>
+
+extern const AP_HAL::HAL& hal;
+
+/// Constructor
+AP_BattMonitor_TattuCAN::AP_BattMonitor_TattuCAN(AP_BattMonitor &mon, AP_BattMonitor::BattMonitor_State &mon_state, AP_BattMonitor_Params &params) :
+    CANSensor("Tattu"),
+    AP_BattMonitor_Backend(mon, mon_state, params)
+{
+    // starts with not healthy
+    _state.healthy = false;
+    register_driver(AP_CANManager::Driver_Type_Tattu);
+}
+
+void AP_BattMonitor_TattuCAN::handle_frame(AP_HAL::CANFrame &frame)
+{
+    if (frame.dlc == 0) {
+        // sanity check for if there's no payload
+        return;
+    }
+
+    const uint8_t payload_len = (frame.dlc - 1); // last payload byte is tail which is not _message data
+    const uint8_t tail_byte = frame.data[frame.dlc-1];
+    const bool start = (tail_byte & TAIL_BYTE_BITS_START_OF_TRANSFER) != 0;
+    const bool end = (tail_byte & TAIL_BYTE_BITS_END_OF_TRANSFER) != 0;
+    //const bool toggle = (tail_byte & TAIL_BYTE_BITS_TOGGLE) != 0;
+    //const bool id = (tail_byte & TAIL_BYTE_TRANSFER_ID_MASK);
+
+
+    if ((start && end) || (_message_offset == 0 && !start)) {
+        // invalid or we're expecting a start and did't get one (re-sync)
+        _message_resync++;
+        return;
+
+    } else if (start) {
+        // start of a frame
+        _message_offset = 0;
+        _message_toggle_expected = 0;
+        
+    } else if ((_message_offset + frame.dlc - 1) > sizeof(_message)) {
+        // buffer overflow, reset
+        // TODO: this only supports 12S batteries. If a 14S is connected, 4 more bytes come in the middle of the packet
+        _message_offset = 0;
+        _message_resync++;
+        return;
+    }
+
+    memcpy(&data[_message_offset], frame.data, payload_len);
+    _message_offset += payload_len;
+    _message_toggle_expected = !_message_toggle_expected;
+
+    if (!end) {
+        return;
+    }
+
+    // message complete!
+    memcpy((uint8_t*)&_message, data, _message_offset);
```

Comment:
Why do we need this memcpy? Why not just have _message and data be a union, and then the only other things you need to store are just the cycle/remaining percentages, which are both currently being read without a semaphore anyway, and could be corrupt readings.